runtime/strings/strings.cpp

```diff
@@ -0,0 +1,291 @@
+	memcpy(&(ret->data[a->b.len]), &(b->data), b->b.len * sizeof(KCHAR));
+	return ret;
+}
+
+int64_t hook_STRING_length(const string * a) {
+	return a->b.len;
+}
+
+string * hook_STRING_chr(const int64_t ord) {
+	auto ret = static_cast<string *>(malloc(sizeof(string) + sizeof(KCHAR)));
+	ret->b.len = 1;
+	ret->data[0] = static_cast<KCHAR>(ord);
+	return ret;
+}
+
+int64_t hook_STRING_ord(const string * input) {
+	if (input->b.len < 1) {
+		return -1;
+	}
+    return static_cast<int64_t>(input->data[0]);
+}
+
+// -1 means take the entire rest of the string.
+string * hook_STRING_substr(const string * input, const int64_t start, int64_t len) {
+	if (len == -1) len = input->b.len;
+	auto minLen = std::max(std::min(len, input->b.len - start), 0L);
+	auto ret = static_cast<string *>(malloc(sizeof(string) + sizeof(KCHAR) * minLen));
+	ret->b.len = minLen;
+	memcpy(&(ret->data), &(input->data[start]), minLen * sizeof(KCHAR));
+	return ret;
+}
+
+int64_t hook_STRING_find(const string * haystack, const string * needle, int64_t pos) {
+	if (pos >= haystack->b.len) {
+		return -1;
+	}
+	auto out = std::search(haystack->data + pos * sizeof(KCHAR), haystack->data + haystack->b.len * sizeof(KCHAR),
+			               needle->data,   needle->data   + needle->b.len * sizeof(KCHAR));
+	int64_t ret = (out - haystack->data) / sizeof(KCHAR);
+	// search returns the end of the range if it is not found, but we want -1 in such a case.
+	return (ret < haystack->b.len)?ret:-1;
+}
+
+int64_t hook_STRING_rfind(const string * haystack, const string * needle, int64_t pos) {
+	// The semantics of rfind position are strange, it is the last position at which
+	// the match can _start_, which means the end of the haystack needs to be pos + len(needle),
+	// or the end of the haystack, if that's less.
+	pos += needle->b.len;
+	if (pos <= needle->b.len) {
+		return -1;
+	}
+	auto end = (pos < haystack->b.len)?pos:haystack->b.len;
+	auto out = std::find_end(&haystack->data[0], &haystack->data[end],
+                  &needle->data[0], &needle->data[needle->b.len]);
+	auto ret = &*out - &haystack->data[0];
+	return (ret < end)?ret:-1;
+}
+
+int64_t hook_STRING_findChar(const string * haystack, const string * needle, int64_t pos) {
+	if (needle->b.len > 1) return -1;
+	return hook_STRING_find(haystack, needle, pos);
+}
+
+int64_t hook_STRING_rfindChar(const string * haystack, const string * needle, int64_t pos) {
+	if (needle->b.len > 1) return -1;
+	return hook_STRING_rfind(haystack, needle, pos);
+}
+
+string * makeString(const KCHAR * input) {
+	auto len = strlen(input);
+	auto ret = static_cast<string *>(malloc(sizeof(string) + len));
+	for (unsigned i = 0; i < len; ++i) {
+		ret->data[i] = input[i];
+	}
+	ret->b.len = len;
+	return ret;
+}
+
+string * hook_STRING_float2string(const double input) {
+	std::stringstream out;
+	out << std::setprecision(14) << input;
+	return makeString(out.str().c_str());
+}
+
+double hook_STRING_string2float(const string * input) {
```

Comment:
this is going to return an mpfr pointer and will call mpfr_set_str, but again, just throw an exception for now. Do fix the signature on all of these functions though.